pragma-node/src/infra/repositories/onchain_repository/history.rs

```diff
@@ -70,3 +103,175 @@ pub async fn get_historical_aggregated_entries(
 
     Ok(raw_entries)
 }
+
+/// Retry to get the onchain historical entries by finding
+/// an alternative route.
+/// TODO: This code is very similar to the one in [entry_repository] ;
+///       once we have proper E2E tests, we should try to merge the code.
+pub async fn retry_with_routing(
+    onchain_pool: &Pool,
+    offchain_pool: &Pool,
+    network: &Network,
+    pair_id: String,
+    timestamp_range: &TimestampRange,
+    chunk_interval: &Interval,
+) -> Result<(Vec<HistoricalEntryRaw>, u32), InfraError> {
+    let [base, quote]: [&str; 2] = pair_id
+        .split('/')
+        .collect::<Vec<_>>()
+        .try_into()
+        .map_err(|_| InfraError::InternalServerError)?;
+
+    let offchain_conn = offchain_pool.get().await.map_err(adapt_infra_error)?;
+    let alternative_currencies = offchain_conn
+        .interact(Currency::get_abstract_all)
+        .await
+        .map_err(adapt_infra_error)?
+        .map_err(adapt_infra_error)?;
+
+    let existing_pairs = get_existing_pairs(onchain_pool, network).await?;
+
+    for alt_currency in alternative_currencies {
+        let base_alt_pair = format!("{}/{}", base, alt_currency);
+        let alt_quote_pair = format!("{}/{}", quote, alt_currency);
+
+        if onchain_pair_exist(&existing_pairs, &base_alt_pair)
+            && onchain_pair_exist(&existing_pairs, &alt_quote_pair)
+        {
+            let base_alt_result = get_historical_entries_and_decimals(
+                onchain_pool,
+                offchain_pool,
+                network,
+                base_alt_pair,
+                timestamp_range,
+                chunk_interval,
+            )
+            .await?;
+            let alt_quote_result = get_historical_entries_and_decimals(
+                onchain_pool,
+                offchain_pool,
+                network,
+                alt_quote_pair,
+                timestamp_range,
+                chunk_interval,
+            )
+            .await?;
+
+            if base_alt_result.0.len() != alt_quote_result.0.len() {
+                continue;
+            }
+
+            return calculate_rebased_prices(base_alt_result, alt_quote_result);
+        }
+    }
+
+    Err(InfraError::RoutingError)
+}
+
+/// Given two vector of entries, compute a new vector containing the routed prices.
+fn calculate_rebased_prices(
+    base_result: (Vec<HistoricalEntryRaw>, u32),
+    quote_result: (Vec<HistoricalEntryRaw>, u32),
+) -> Result<(Vec<HistoricalEntryRaw>, u32), InfraError> {
```

Comment:
too risky to make the function we already have generic on some trait and implement that trait on both our types?

Reply:
I'd rather make this when we have tests for `get_entry` as this will impact it also

Reply:
let's create issue then

Reply:
Added here: #118